L1Trigger/L1TZDC/plugins/L1TZDCProducer.cc

```diff
@@ -0,0 +1,260 @@
+
+#include "L1Trigger/L1TCalorimeter/interface/CaloParamsHelper.h"
+#include "CondFormats/DataRecord/interface/L1TCaloParamsRcd.h"
+#include "CondFormats/DataRecord/interface/L1TCaloParamsO2ORcd.h"
+
+#include "DataFormats/L1TCalorimeter/interface/CaloTower.h"
+#include "DataFormats/L1Trigger/interface/EGamma.h"
+#include "DataFormats/L1Trigger/interface/Tau.h"
+#include "DataFormats/L1Trigger/interface/Jet.h"
+#include "DataFormats/L1Trigger/interface/EtSum.h"
+
+#include "DataFormats/HcalDigi/interface/HcalDigiCollections.h"
+
+//
+// class declaration
+//
+
+using namespace l1t;
+
+class L1TZDCProducer : public edm::stream::EDProducer<> {
+public:
+  explicit L1TZDCProducer(const edm::ParameterSet& ps);
+  ~L1TZDCProducer() override = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  //  void endRun(edm::Run const&, edm::EventSetup const&) override;
+
+  int zdcLUTIndexHelper(int iDetPos, int iBXPos);
+
+  // ----------member data ---------------------------
+
+  // input tokens
+  //Add the ZDC token, candidateToken for caloParamsHelper
+  edm::EDGetTokenT<QIE10DigiCollection> m_zdcToken;
+  edm::ESGetToken<CaloParams, L1TCaloParamsRcd> m_candidateToken;
+
+  //input ints
+  int bxFirst_;
+  int bxLast_;
+  int sampleToCenterBX_;
+
+  // put tokens
+  edm::EDPutTokenT<EtSumBxCollection> m_etToken;
+
+  //Following the L1TStage2Layer2Producer
+  std::unique_ptr<CaloParamsHelper> m_params;
+  int m_scaleFactor;
+};
+
+L1TZDCProducer::L1TZDCProducer(const edm::ParameterSet& ps) {
+  // register what you produce
+  m_etToken = produces<EtSumBxCollection>("zdcEtSums");
+
+  // register what you consume and keep token for later access:
+  m_zdcToken = consumes<QIE10DigiCollection>(ps.getParameter<edm::InputTag>("zdcToken"));
+  m_candidateToken = esConsumes<CaloParams, L1TCaloParamsRcd, edm::Transition::BeginRun>();
+
+  bxFirst_ = ps.getParameter<int>("bxFirst");
+  bxLast_ = ps.getParameter<int>("bxLast");
+  sampleToCenterBX_ = ps.getParameter<int>("sampleToCenterBX");
+}
+
+// ------------ method called to produce the data  ------------
+void L1TZDCProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  using namespace l1t;
+
+  LogDebug("l1t|stage 2") << "L1TZDCProducer::produce function called..." << std::endl;
+
+  // reduced collection to be emplaced in output
+  EtSumBxCollection etsumsReduced(0, bxFirst_, bxLast_);
+
+  //inputs
+  Handle<QIE10DigiCollection> zdcDigiCollection;
+  iEvent.getByToken(m_zdcToken, zdcDigiCollection);
+
+  //Produce ZDC EtSums IF input zdcDigiCollection is valid
+  if (zdcDigiCollection.isValid()) {
+    //In lieu of bxFirst, bxLast, use the number of the timeslice samples
+    const QIE10DataFrame& frametest = (*zdcDigiCollection)[0];
+    int nSamples = frametest.samples();
```

Comment:
what is the relationship between `bxFirst_`, bxLast_`, `sampleToCenterBX_`, and the ZDC digis' `samples()` and `presamples()` ?

Reply:
sampleToCenterBX_ is just user-designated sample that should be treated as BX=0 by the emulator. bxFirst_ and bxLast_ tells the L1TZDCProducer how many samples around that selected center to retain (so in the defaults, sample 4 is designated 0, and -2 to +2 around it retained; in sample terms, 2-6).

There is no in-code relationship other than the fact that we loop over the nSamples (presamples used at no point in code) but I suppose in a sense the maximum value for firstBX_ allowed should be set to -presamples and the maximum for lastBx_ samples - 1 - presamples.

I should note that the QIE10DataFrame (like the ZDC digis collection you flag) actually also has samples and presamples. While the code makes use of samples for processing the QIE10DigiCollection, I had not used the presamples so I did a quick check - samples in files we've analyzed is always 10 and presamples always 4. the latter of which matches our default centered bx sample of position 4 (0-3 preceding). This also corresponds to where the ZCDEtSum value peaks, which I think ought be bx 0.

^Let me know if this was the information you were looking for or if I misunderstood.

Reply:
I think I understand - the minimal constraint I think would be what I describe above on firstBX_ and lastBX_, in addition to checking requested BX wouldnt correspond to nSamples index of < 0 or >= nSamples.

I looked into how the bx is assigned for the caloTowers collection, and at least when coming from L1TCaloLayer1 (I think primary producer here), it's just set to 0 or the hit bx; see at the link:

https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TCaloLayer1/plugins/L1TCaloLayer1.cc#L249

Based on some limited conversations in the past w/ ZDC experts, I think there can be timing issues w/ the ZDC requiring we don't fix it exactly to presamples, but perhaps +-1 around presamples? And combine it with the protections on firstBX_ and lastBX_ mentioned above? Could alternatively have some default configuration that uses nPresamples as BX = 0, but could be turned off for this existing selection mode w/ a boolean. Let me know what makes sense to you

Reply:
I'm afraid that suggesting or commenting on the proper constraints goes beyond my knowledge of HCAL, ZDC and L1T :-(

Reply:
Ok, understood - I will discuss with experts then - the minimal constraints can go in to at least protect the use from selecting invalid nSamples ranges with the first and last bx options, and I'll complete the open items in the interim.

Reply:
Implemented bare-bones protections as discussed in latest commit - will discuss with ZDC experts more appropriate protections